rdflib/collection.py

```diff
@@ -203,30 +219,58 @@ def append(self, item):
         >>> len([i for i in links if (i,RDF.rest, RDF.nil) in g])
         1
 
+        The following test can take a long time if append is not implemented correctly
+        >>> from rdflib.graph import Namespace
+        >>> from rdflib.collection import Collection
+        >>> g = Graph()
+        >>> EX = Namespace("http://example.org/ex#")
+        >>> c = Collection(g, BNode(), [EX[str(i)] for i in range(1,10000)])
+        >>> len(c)
+        9999
+        >>> c.append(EX['new'])
+        >>> str(c[-1])
+        'http://example.org/ex#1'
+        >>> del c[9999]
+        >>> len(c)
+        9999
+        >>> str(c[9998])
+        'http://example.org/ex#9999'
+        >>> c.clear()
+        >>> c.append(EX['1'])
+        >>> len(c)
+        1
         """
-        container = self.uri
         graph = self.graph
-        # iterate to the end of the linked list
-        rest = graph.value(container, RDF.rest)
-        while rest:
-            if rest == RDF.nil:
-                # the end, append to the end of the linked list
-                node = BNode()
-                graph.set((container, RDF.rest, node))
-                container = node
-                break
-            else:
-                # move down one link
-                if container != self.uri:
-                    rest = graph.value(rest, RDF.rest)
-                if not rest == RDF.nil:
-                    container = rest
-        graph.add((container, RDF.first, item))
-        graph.add((container, RDF.rest, RDF.nil))
+        if graph.value(self._end, RDF.rest) == RDF.nil:
```

Comment:
I prefer `(self._end, RDF.rest, RDF.nil) in graph`, which may also be slightly faster?